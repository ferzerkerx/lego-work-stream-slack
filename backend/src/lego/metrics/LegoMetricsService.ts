import { LegoSelectMessage } from '../LegoSelectMessage';
import { LegoMetricsCalculator } from './LegoMetricsCalculator';
import { Metrics } from './Metrics';
import { LegoMetricsService, LegoSelectMessageRepository } from '../Types';

export class LegoMetricsServiceImpl implements LegoMetricsService {
  constructor(private repository: LegoSelectMessageRepository) {}

  metricsForConfig(config: MetricsConfiguration): Promise<Metrics> {
    return this.repository
      .findMessagesBy(config)
      .then((legoSelectMessages: LegoSelectMessage[]) => {
        return LegoMetricsCalculator.calculate(legoSelectMessages, config);
      });
  }
}
