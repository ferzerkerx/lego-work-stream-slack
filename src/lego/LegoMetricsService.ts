import { Storage } from 'botkit';
import { LegoSelectMessage } from './LegoSelectMessage';
import { DateUtils } from '../DateUtils';
import { LegoMetricsCalculator, MetricEntry } from './LegoMetricsCalculator';

export class LegoMetricsService {
  static metricsForConfig(
    storage: Storage<LegoSelectMessage>,
    config: MetricsConfiguration
  ): Promise<MetricEntry> {
    const messagesPerDatePromises: Promise<
      LegoSelectMessage[]
    >[] = this.toDatesArray(config.startDate, config.endDate).map(theDate =>
      this.findMessagesBy(storage, theDate)
    );

    return Promise.all(messagesPerDatePromises).then(
      (allMessages: LegoSelectMessage[][]) => {
        const legoSelectMessages: LegoSelectMessage[] = allMessages.reduce(
          (
            accumulator: LegoSelectMessage[],
            currentValue: LegoSelectMessage[]
          ) => accumulator.concat(currentValue)
        );
        return LegoMetricsCalculator.calculate(
          legoSelectMessages,
          this.toDatesArray(
            config.startDate,
            config.endDate,
            config.frequencyInDays
          )
        );
      }
    );
  }

  private static findMessagesBy(
    storage: Storage<LegoSelectMessage>,
    theDate: Date
  ): Promise<LegoSelectMessage[]> {
    // @ts-ignore
    return storage.lego_messages
      .find({ date: theDate })
      .catch(e => this.defaultErrorHandling(e));
  }

  static toDatesArray(
    startDate: Date,
    endDate?: Date,
    interval: number = 1
  ): Array<Date> {
    const theDates = [startDate];
    if (endDate) {
      if (endDate < startDate) {
        throw Error('End date should be bigger');
      }
      let currentDate = new Date(startDate);
      while (currentDate < endDate) {
        currentDate = DateUtils.add(currentDate, interval);
        theDates.push(currentDate);
      }
    }
    return theDates;
  }

  private static defaultErrorHandling(err: Error): void {
    console.error(err);
  }
}
