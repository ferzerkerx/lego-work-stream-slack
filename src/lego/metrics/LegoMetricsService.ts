import { Storage } from 'botkit';
import { LegoSelectMessage } from '../LegoSelectMessage';
import { DateUtils } from '../../DateUtils';
import { LegoMetricsCalculator } from './LegoMetricsCalculator';
import { Metrics } from './Metrics';

export class LegoMetricsService {
  static metricsForConfig(
    storage: Storage<LegoSelectMessage>,
    config: MetricsConfiguration
  ): Promise<Metrics> {
    const messagesPerDatePromises: Promise<
      LegoSelectMessage[]
    >[] = DateUtils.toDatesArray(config.startDate, config.endDate).map(
      theDate => this.findMessagesBy(storage, theDate)
    );

    return Promise.all(messagesPerDatePromises).then(
      (allMessages: LegoSelectMessage[][]) => {
        const legoSelectMessages: LegoSelectMessage[] = allMessages.reduce(
          (
            accumulator: LegoSelectMessage[],
            currentValue: LegoSelectMessage[]
          ) => accumulator.concat(currentValue)
        );
        return LegoMetricsCalculator.calculate(legoSelectMessages, config);
      }
    );
  }

  private static findMessagesBy(
    storage: Storage<LegoSelectMessage>,
    theDate: Date
  ): Promise<LegoSelectMessage[]> {
    // @ts-ignore
    return storage.lego_messages
      .find({ date: theDate })
      .catch(e => this.defaultErrorHandling(e));
  }

  private static defaultErrorHandling(err: Error): void {
    console.error(err);
  }
}
