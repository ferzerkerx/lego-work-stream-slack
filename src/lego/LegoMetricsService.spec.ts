import { LegoMetricsService } from './LegoMetricsService';

describe('LegoMetricsService', () => {
  test('toDatesArray startdate only', () => {
    const entry: any = LegoMetricsService.toDatesArray(new Date('2018-10-10'));
    expect(entry).toMatchSnapshot();
  });

  test('toDatesArray startdate and endDate', () => {
    const entry: any = LegoMetricsService.toDatesArray(
      new Date('2018-10-10'),
      new Date('2018-10-15')
    );
    expect(entry).toMatchSnapshot();
  });
});
