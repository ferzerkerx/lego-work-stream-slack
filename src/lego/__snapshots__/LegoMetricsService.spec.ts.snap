// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoMetricsService toDatesArray startdate and endDate 1`] = `
Array [
  2018-10-10T00:00:00.000Z,
  2018-10-11T00:00:00.000Z,
  2018-10-12T00:00:00.000Z,
  2018-10-13T00:00:00.000Z,
  2018-10-14T00:00:00.000Z,
  2018-10-15T00:00:00.000Z,
]
`;

exports[`LegoMetricsService toDatesArray startdate only 1`] = `
Array [
  2018-10-10T00:00:00.000Z,
]
`;
